### PR TITLE
Improve properties.mako.rs file structure

### PR DESCRIPTION
**Pull request by perlun**
Background: when I started hacking on Servo some week(s) ago, I quite quickly came in contact with the `components/style/properties.mako.rs` file. I was surprised to see that it was so large (7000 lines of code). In my book, that's too much; there must clearly be some way to structure things in a more reasonable way, I feel.

I am by no means a Python nor a Mako expert (more like: I've never used it before experiencing it in Servo), but I started googling about it yesterday and saw that Mako has an `%include` system that could be used for this kind of stuff. Said and done, I started playing around with it and managed to get something that compiles today - I have used this approach now for extracting a set of properties from the big file.

I realize this is a significant change to the structure here, and this PR serves more as a starting point for discussion than anything else. There may be other, superior ways of achieving the same end result (i.e. splitting the file into smaller parts), better timing for trying to land stuff like this etc etc. Having that said, I find the way I chose here _pretty decent_ to be honest. The `properties.mako.rs` is down from ~7000 LoC to about 5800, by extracting 24 properties into separate files. There are about 70 properties left (of which some hard more difficult, because of scoping etc - they use local methods declared in the `properties.mako.rs` file which are un-callable from include files (have to use `%def` but it didn't seem trivial when I started looking at that, so skipped those properties for now). So, all in all, I think we should be able to bring down this 7000 line file to maybe 2-3000 lines or something quite easily, which would IMHO be a great start.

Opinions/flames/etc? :sunglasses:

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10586)

### CODE REVIEW

File: components/style/build.rs [hunk 1]
```diff
@@ -43,19 +43,7 @@ fn main() {
...
-import sys
-from mako.template import Template
-from mako import exceptions
-try:
-    template = Template(open(os.environ['TEMPLATE'], 'rb').read(), input_encoding='utf8')
-    print(template.render(PRODUCT=os.environ['PRODUCT']).encode('utf8'))
-except:
-    sys.stderr.write(exceptions.text_error_template().render().encode('utf8'))
-    sys.exit(1)
-"#)
```
Comment (perlun) [inline components/style/build.rs]: This change is not _mandatory_ in the PR, but it felt natural in line with the thinking to "avoid multiple languages in the same file" (can't find the reference now...). Maintaining Python code is also easier when it has a proper `.py` suffix etc.

**Comment by KiChjang** [comment]
Oh yes, this is definitely a nice change! I believe the idea was to actually do away with mako altogether and use native Rust macros to generate code, but this is an interim state that is still desirable, given that `properties.mako.rs` is bloated as it is right now.

**Comment by perlun** [comment]
Yeah, I also thought about native Rust macros (which would be the natural approach if possible). But, maybe that's more a long-term goal and trying to "clean up the current mess" can be the short-term fix.

File: components/style/build.rs (hunk 1 above)
Comment (bholley) [inline components/style/build.rs]: Can you fix this in geckolib too?

**Comment by bholley** [comment]
Yeah, I would be pretty opposed to getting rid of the mako stuff right now. It's extremely flexible and very understandable for non-experts, both of which are crucial for the Stylo work right now. Everybody complains about it, but I haven't heard a cogent plan for how we would do all these things concisely via the AST, or why it would be better.

As for this PR (splitting out properties into separate files): there are pros and cons, but I think it's probably a net win (assuming we follow through with it). As long as I can still import all the data structures from geckolib, I'm happy. I'll let @SimonSapin decide though. :-)

**Comment by SimonSapin** [comment]
@bholley Something based on a syntax extension would allow us to have Rust error messages use line numbers from source files rather than a generated file. But maybe https://github.com/rust-lang/rfcs/pull/1573 (source maps) will eventually be another way to do that.

I agree with the general idea of this PR: `properties.mako.rs` is too big, and replacing Mako is not gonna happen soon. But I’d like a more systematic way to decide what goes into what files. One file per property is probably too many files. Maybe one file per style struct?

I think we can separate even more data definition from code generation. I’ll look in more details tomorrow.

**Comment by KiChjang** [comment]
```
$ ./mach test-unit
Traceback (most recent call last):
  File "/home/travis/build/servo/servo/components/style/list_properties.py", line 16, in <module>
    template.render(PRODUCT='servo')
  File "/home/travis/build/servo/servo/components/style/Mako-0.9.1.zip/mako/template.py", line 443, in render
  File "/home/travis/build/servo/servo/components/style/Mako-0.9.1.zip/mako/runtime.py", line 807, in _render
  File "/home/travis/build/servo/servo/components/style/Mako-0.9.1.zip/mako/runtime.py", line 839, in _render_context
  File "/home/travis/build/servo/servo/components/style/Mako-0.9.1.zip/mako/runtime.py", line 865, in _exec_template
  File "_home_travis_build_servo_servo_components_style_properties_mako_rs", line 225, in render_body
  File "/home/travis/build/servo/servo/components/style/Mako-0.9.1.zip/mako/runtime.py", line 730, in _include_file
  File "/home/travis/build/servo/servo/components/style/Mako-0.9.1.zip/mako/runtime.py", line 770, in _lookup_template
mako.exceptions.TemplateLookupException: Template '/home/travis/build/servo/servo/components/style/properties.mako.rs' has no TemplateLookup associated
Error running mach:
    ['test-unit']
The error occurred in code that was called by the mach command. This is either
a bug in the called code itself or in the way that mach is calling it.
You should consider filing a bug for this issue.
If filing a bug, please include the full output of mach, including this error
message.
The details of the failure are as follows:
CalledProcessError: Command '['/home/travis/virtualenv/python2.7.10/bin/python2.7', u'/home/travis/build/servo/servo/components/style/list_properties.py']' returned non-zero exit status 1
  File "/home/travis/build/servo/servo/python/servo/testing_commands.py", line 146, in test_unit
    path.join(self.context.topdir, "components", "style", "list_properties.py")
  File "/opt/python/2.7.10/lib/python2.7/subprocess.py", line 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
The command "./mach test-unit" exited with 1.
```

**Comment by perlun** [comment]
Thanks to you all for the feedback thus far!

- @bholley - As for extracting the Mako scripts to `.py` for geckolib, yes sure - I can fix this.
- @KiChjang - the unit test failures; thanks for pointing that out, will look into that as well.

Regarding the structure (how to split, one file per property or not), that's definitely one of the harder questions. I don't really have a full understanding of the structure here yet; one file per property didn't feel completely unreasonable (yes, it gives us many files but OTOH we can put the property name in the file name making it quite easy to find what you're looking for). But absolutely, if we can agree on a better, more balanced structure, I'm all open for suggestions. (@SimonSapin, to make it even clearer - please suggest a few of the file names so I get your general idea of the overall picture)

**Comment by perlun** [comment]
FWIW, here are the file lengths of the new files introduced in the PR. I don't know what the recommended "reasonable" maximum class/code file length is in Rust-land? In the domain I normally work (Ruby), 100 lines is considered a good maximum for a class. Do we have a recommendation yet in the Rust community?

```
$ wc -l properties/*
      61 properties/background-image.mako.rs
     114 properties/background-position.mako.rs
     137 properties/background-size.mako.rs
      39 properties/border-widths.mako.rs
      37 properties/color.mako.rs
     164 properties/content.mako.rs
      70 properties/counter-increment.mako.rs
      10 properties/counter-reset.mako.rs
      20 properties/float.mako.rs
     110 properties/font-family.mako.rs
      66 properties/font-size.mako.rs
     119 properties/font-weight.mako.rs
      63 properties/letter-spacing.mako.rs
      93 properties/line-height.mako.rs
      66 properties/list-style-image.mako.rs
      15 properties/outline-style.mako.rs
      33 properties/outline-width.mako.rs
      39 properties/overflow-y.mako.rs
      67 properties/quotes.mako.rs
      17 properties/servo-display-for-hypothetical-box.mako.rs
      53 properties/text-align.mako.rs
      87 properties/text-decoration.mako.rs
      63 properties/word-spacing.mako.rs
      47 properties/z-index.mako.rs
    1590 total
```

**Comment by SimonSapin** [comment]
I don’t think there is such a convention in Rust. (FWIW 100 lines seems very small to me, I don’t understand why Rubyists considers that a maximum.)

**Comment by bholley** [comment]
Yeah, doing it per-style-struct seems sensible IMO.

**Comment by perlun** [comment]
Yeah, I know 100 lines is a bit extreme :smile: but it actually works quite well in the Ruby case; it's a pretty expressive language that has good mechanics in place to help keeping classes short (then again, sometimes it's just not possible). I think the general idea of having _some_ rule or recommendation within a language community makes sense; it gives new developers also a feeling of "what is reasonable" - to indicate signs of code smell, and when stuff should be acted upon. Even Microsoft had this in the C# days I think (googled it a bit couldn't find the specifics), although their rule was probably a lot more than 100 LoC... :smile: 

----

Anyway, Rust is not Ruby and perhaps we shouldn't get sidetracked too much. I looked a bit at the file again today, trying to understand your POV better @bholley and @SimonSapin - you were speaking about one file per style struct. Are you talking about these modules?

```rust
pub mod longhands
pub mod shorthands
pub mod style_struct_traits
pub mod style_structs
```

I.e. to split it in four pieces instead. Was that what you were suggesting? Or was it something else?

**Comment by jdm** [comment]
I believe the suggestion was to split it at the level of each `new_style_struct` invocation.

**Comment by bholley** [comment]
@jdm is correct. Note that the current scheme jumps around a bit with the switch_to_style_struct stuff. That would presumably be consolidated.

**Comment by perlun** [comment]
Alright, that makes sense. Will give it a try and carve something out based on these suggestions (but probably only for a small subset of the file first to ensure I get it right). Thanks.

**Comment by SimonSapin** [comment]
I believe I originally introduced `switch_to_style_struct` to keep properties roughly in the order of CSS 2.1 chapters, but that is not really important. We can get rid of it by moving property definitions around (which is happening anyway if they move to other files).

**Comment by SimonSapin** [comment]
I had some ideas around this but they were a bit complex to explain, so I went ahead and did it: https://github.com/servo/servo/compare/split-mako

It does not reduce (much) the size of `properties.mako.rs` yet, but I think it’s a good base to have a nicer split between data definition and code generation.

I’ll work on this some more and send a PR, which would replace this one. Sorry for kinda stealing this from you @perlun !

**Comment by perlun** [comment]
Ouch man, that hurts... :smiley: I guess I should have checked my inbox _before_ spending like the 3rd night or something on this, trying to accommodate to the suggestions given by you... :wink: It was really, really hard to be able to get the include stuff working but I **finally** managed to get it done. It feels good, but I had to learn a bit more about Mako than I'd really have wished for to be able to get it done.

So, here goes: what I have at the moment is pushed to this branch. Please check it out if you like. As stated, I wasn't aware of the work you'd done on this until now, when I pushed them up and read your message. :innocent: 

Having that said: I don't take pride in _forcing_ my way through here. Your changes are really good, from what I understand of them: trying to structure things in a bit better way. I think I have most of my changes well under control, so even though it will be awful to having to rebase it to merge with your stuff... well... Either that way, or you allow this stuff through first, and then rebase your branch on top off that once it's landed in `master`. Given that you _did_ in fact snatch my PR, it wouldn't be entirely insane to suggest that... :wink: 

(to be very clear, humour aside: no hard feelings towards you whatsoever, really. I do appreciate your work in this area, especially since I know how awkward it can be to work with the Python/Mako stuff. Yes, I know it very very well indeed. :stuck_out_tongue:)

(final PS: I realize you perhaps thought the PR was going stale or something. The reason why it took so long time was simply that it was very hard to get progress in the way you suggested.)

File: components/style/list_properties.py [hunk 2]
```diff
@@ -10,9 +10,14 @@
...
 sys.path.insert(0, os.path.join(style, "Mako-0.9.1.zip"))
+
+from mako.lookup import TemplateLookup
 from mako.template import Template
 
-template = Template(filename=os.path.join(style, "properties.mako.rs"), input_encoding='utf8')
+lookup = TemplateLookup(directories=[style])
+template = Template(filename=os.path.join(style, "properties.mako.rs"),
+                    input_encoding='utf8',
+                    lookup=lookup)
```
Comment (perlun) [inline components/style/list_properties.py]: These changes are needed to be able to use %include. It was significantly harder to be able to use it from `list_properties.py`, so in that sense your (Simon) change in the other branch where you refactor `list_properties.py` away is kind of nice.

File: components/style/properties.mako.rs [hunk 3]
```diff
@@ -128,38 +143,55 @@ class StyleStruct(object):
...
-CONFIG = {}
-
-def set_product(p):
-    global CONFIG
-    CONFIG['product'] = p
+%>
 
-def new_style_struct(name, is_inherited, gecko_name=None, additional_methods=None):
-    global THIS_STYLE_STRUCT
+<%def name="new_style_struct(name, is_inherited, gecko_name=None, additional_methods=None)">
```
Comment (perlun) [inline components/style/properties.mako.rs]: Very much the key to getting stuff working in the `%include`d files is to move all stuff that needs to be callable into proper `%def` methods.

**Comment by SimonSapin** [comment]
No, I didn’t think this PR was stale and did realize that you might still be working on this. I should have communicated better from the start, sorry. I’ll look tomorrow at your latest changes and try to integrate both branches.

**Comment by perlun** [comment]
Sounds great! G'night for now.

**Comment by perlun** [comment]
(I noted there were a bunch of linting issues. Feel free to fork my branch and fix them if you like, if you want to speed the integration of these changes. Otherwise I'll try to sort them out tomorrow. I'd unfortunately only verified that this built, that geckolib built, that the unit tests ran... but hadn't verified the tidy checks before pushing...)

**Comment by perlun** [comment]
Tidiness checks should now be OK. Please check and let me know if we can merge as-is or if you want it squashed up a bit.

**Comment by perlun** [comment]
@SimonSapin - any progress on this one? I'm eager to continue moving stuff out of the monster file... :smile:

**Comment by SimonSapin** [comment]
I’m in the middle of responding to 34 review comments on another PR. This is next :)

**Comment by perlun** [comment]
> I’m in the middle of responding to 34 review comments on another PR.

Oh wow! I don't envy you. :wink:

**Comment by perlun** [comment]
Closing this now Simon - I have a new PR coming up, but I'm unable to submit it until #10749 is landed (since I branched off of that branch; my branch now includes all your commits until I'm able to rebase it on top of master once homu/bors/buildbots have done their work). Thanks a lot for what you've done so far! Sorry if I sounded a bit rude the other night; it was kind of annoying yes, but... I think the end result is a much nicer codebase in these parts, which is something which we will _all_ benefit from in the long run.

**Pull request closed without merging**